go/database/mpt/io/parallel_visit.go

```diff
@@ -0,0 +1,429 @@
+	}
+}
+
+func (p *position) compare(b *position) int {
+	if p == b {
+		return 0
+	}
+	if p == nil {
+		return -1
+	}
+	if b == nil {
+		return 1
+	}
+
+	// make sure a is the shorter one
+	if p.len > b.len {
+		return b.compare(p) * -1
+	}
+
+	// reduce the length of b to match a
+	bIsLonger := p.len < b.len
+	for p.len < b.len {
+		b = b.parent
+	}
+
+	// compare the common part
+	prefixResult := p._compare(b)
+	if prefixResult != 0 {
+		return prefixResult
+	}
+	if bIsLonger {
+		return -1
+	}
+	return 0
+}
+
+func (p *position) _compare(b *position) int {
+	if p == b {
+		return 0
+	}
+	prefixResult := p.parent._compare(b.parent)
+	if prefixResult != 0 {
+		return prefixResult
+	}
+	if p.pos < b.pos {
+		return -1
+	}
+	if p.pos > b.pos {
+		return 1
+	}
+	return 0
+}
+
+// ----------------------------------------------------------------------------
+//                               NodeSource
+// ----------------------------------------------------------------------------
+
+type nodeSourceFactory struct {
+	directory string
+}
+
+func (f *nodeSourceFactory) Open() (NodeSource, error) {
+	accounts, err := openSource[mpt.AccountNode](f.directory, "accounts", mpt.AccountNodeWithPathLengthEncoderWithNodeHash{})
+	if err != nil {
+		return nil, err
+	}
+	branches, err := openSource[mpt.BranchNode](f.directory, "branches", mpt.BranchNodeEncoderWithNodeHash{})
+	if err != nil {
+		return nil, err
+	}
+	extensions, err := openSource[mpt.ExtensionNode](f.directory, "extensions", mpt.ExtensionNodeEncoderWithNodeHash{})
+	if err != nil {
+		return nil, err
+	}
+	values, err := openSource[mpt.ValueNode](f.directory, "values", mpt.ValueNodeWithPathLengthEncoderWithNodeHash{})
+	if err != nil {
+		return nil, err
+	}
+	return &nodeSource{
+		accounts:   accounts,
+		branches:   branches,
+		extensions: extensions,
+		values:     values,
+	}, nil
+}
+
+type nodeSource struct {
+	accounts   source[mpt.AccountNode]
+	branches   source[mpt.BranchNode]
+	extensions source[mpt.ExtensionNode]
+	values     source[mpt.ValueNode]
+}
+
+func (s *nodeSource) Get(id mpt.NodeId) (mpt.Node, error) {
+	if id.IsEmpty() {
+		return mpt.EmptyNode{}, nil
+	}
+	if id.IsAccount() {
+		res, err := s.accounts.Get(id)
+		return &res, err
+	}
+	if id.IsBranch() {
+		res, err := s.branches.Get(id)
+		return &res, err
+	}
+	if id.IsExtension() {
+		res, err := s.extensions.Get(id)
+		return &res, err
+	}
+	if id.IsValue() {
+		res, err := s.values.Get(id)
+		return &res, err
+	}
+	return nil, fmt.Errorf("unknown node type: %v", id)
+}
+
+func (s *nodeSource) Close() error {
+	return errors.Join(
+		s.accounts.Close(),
+		s.branches.Close(),
+		s.extensions.Close(),
+		s.values.Close(),
+	)
+}
+
+type source[T any] struct {
+	file    utils.OsFile
+	encoder stock.ValueEncoder[T]
+	buffer  []byte
+}
+
+func openSource[T any](directory, name string, encoder stock.ValueEncoder[T]) (source[T], error) {
+	file, err := os.Open(fmt.Sprintf("%s/%s/values.dat", directory, name))
```

Comment:
If we want to play it safe, the `openSource` function should check whether the directory contains a `stock` with the proper value and index type. This would be easiest by reusing code from the file stock.

Probably this code for the "source" should be moved to the file stock package and renamed to something like a "StockReader" or something like that ...

Reply:
reworked this way